src/lib/comms/sol-bluetooth.c

```diff
@@ -0,0 +1,200 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+
+#include <sol-bluetooth.h>
+#include <sol-buffer.h>
+#include <sol-log.h>
+#include <sol-str-table.h>
+#include <sol-util.h>
+
+#define BASE_UUID { .type = SOL_BT_UUID_TYPE_128, \
+                    .val128 = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, \
+                                0x80, 0x00, 0x00, 0x80, 0x5F, 0x9B, 0x34, 0xFB, } \
+}
+
+static int
+string_to_uuid128(struct sol_bt_uuid *uuid, const char *str)
+{
+    uint32_t data0, data4;
+    uint16_t data1, data2, data3, data5;
+    uint8_t *val = uuid->val128;
+
+    if (sscanf(str, "%08x-%04hx-%04hx-%04hx-%08x%04hx",
+        &data0, &data1, &data2,
+        &data3, &data4, &data5) != 6)
+        return -EINVAL;
+
+    data0 = sol_util_be32_to_cpu(data0);
+    data1 = sol_util_be16_to_cpu(data1);
+    data2 = sol_util_be16_to_cpu(data2);
+    data3 = sol_util_be16_to_cpu(data3);
+    data4 = sol_util_be32_to_cpu(data4);
+    data5 = sol_util_be16_to_cpu(data5);
+
+    memcpy(&val[0], &data0, 4);
+    memcpy(&val[4], &data1, 2);
+    memcpy(&val[6], &data2, 2);
+    memcpy(&val[8], &data3, 2);
+    memcpy(&val[10], &data4, 4);
+    memcpy(&val[14], &data5, 2);
+
+    return 0;
+}
+
+static void
+uuid16_to_uuid128(const struct sol_bt_uuid *u16, struct sol_bt_uuid *u128)
+{
+    uint16_t val;
+
+    val = sol_util_cpu_to_be16(u16->val16);
+    memcpy(&u128->val128[2], &val, sizeof(val));
+}
+
+static void
+uuid32_to_uuid128(const struct sol_bt_uuid *u32, struct sol_bt_uuid *u128)
+{
+    uint32_t val;
+
+    val = sol_util_cpu_to_be32(u32->val32);
+    memcpy(&u128->val128[0], &val, sizeof(val));
+}
+
+static void
+uuid_to_uuid128(const struct sol_bt_uuid *u, struct sol_bt_uuid *u128)
+{
+    switch (u->type) {
+    case SOL_BT_UUID_TYPE_128:
+        *u128 = *u;
+        break;
+    case SOL_BT_UUID_TYPE_32:
+        uuid32_to_uuid128(u, u128);
+        break;
+    case SOL_BT_UUID_TYPE_16:
+        uuid16_to_uuid128(u, u128);
+        break;
+    }
+}
+
+SOL_API int
+sol_bt_uuid_from_str(struct sol_bt_uuid *uuid, const struct sol_str_slice *str)
+{
+    int r;
+
+    SOL_NULL_CHECK(str, -EINVAL);
+    SOL_NULL_CHECK(uuid, -EINVAL);
+
+    switch (str->len) {
+    case 4:
+        uuid->type = SOL_BT_UUID_TYPE_16;
+        uuid->val16 = strtoul(str->data, NULL, 16);
+        break;
+    case 8:
+        uuid->type = SOL_BT_UUID_TYPE_32;
+        uuid->val32 = strtoull(str->data, NULL, 16);
```

Comment:
On 29-04-2016 11:19, Gustavo Lima Chaves wrote:
> In src/lib/comms/sol-bluetooth.c
> <https://github.com/solettaproject/soletta/pull/1948#discussion_r61584488>:
>
>> +SOL_API int
>> +sol_bt_uuid_from_str(struct sol_bt_uuid *uuid, const struct sol_str_slice *str)
>> +{
>> +    int r;
>> +
>> +    SOL_NULL_CHECK(str, -EINVAL);
>> +    SOL_NULL_CHECK(uuid, -EINVAL);
>> +
>> +    switch (str->len) {
>> +    case 4:
>> +        uuid->type = SOL_BT_UUID_TYPE_16;
>> +        uuid->val16 = strtoul(str->data, NULL, 16);
>> +        break;
>> +    case 8:
>> +        uuid->type = SOL_BT_UUID_TYPE_32;
>> +        uuid->val32 = strtoull(str->data, NULL, 16);
>
> |sol_util_strtoul()| again

`sol_util_strtoul` only advantage is being able to work with 
`sol_str_slice`, isn't it?

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/solettaproject/soletta/pull/1948/files/5c5ef9bce1fd2b5299cd656d22d675d449ea2693..83b1cca5b56becd0ab944689ae28668b210c1d84#r61584488>
>